### PR TITLE
TestServer refactor

### PR DESCRIPTION
**Pull request by natanlao**
The commits can be reviewed individually. (That said, the last one is big and messy - this ticket was underspecified, and as a result, this work kind of spiraled into something a lot bigger than it should have. My fault.)

### CODE REVIEW

File: python/test/test_server.py [hunk 1]
```diff
@@ -686,22 +407,13 @@ class TestServerWithLocalClient(TestServer):
...
-        cls._server_process = p
-
-        local_client = Client(SERVER_URL, local=True)
-
-        # setup logging
-        root = logging.getLogger()
-        root.setLevel(logging.ERROR)
-        logging.captureWarnings(True)
-
+        # Set up the client
```
Comment (achave11) [inline python/test/test_server.py]: Nit: Single line comment.

File: python/test/test_server.py [hunk 2]
```diff
@@ -686,22 +407,13 @@ class TestServerWithLocalClient(TestServer):
     the same tests is a little overkill but they're fast enough that it
     really doesn't make a difference at all.
     """
-
     @classmethod
     def setUpClass(cls):
         # Start a test server
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment.

File: python/test/test_server.py (hunk 1 above)
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment.

File: python/test/test_server.py [hunk 3]
```diff
@@ -664,13 +386,12 @@ def test_schema_required(self):
         CreateDataObjectRequest = self._models.get_model('CreateDataObjectRequest')
         DataObject = self._models.get_model('CreateDataObjectRequest')
         # Missing the `id` parameter
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment.

File: python/test/test_server.py [hunk 4]
```diff
@@ -1,660 +1,382 @@
...
-            page_token=list_request.page_token).result()
+        logger.info(list_response.next_page_token)
+        logger.info(ids)
+
+        list_response = self.request('ListDataBundles', page_size=10, page_token=list_response.next_page_token)
         ids = [x['id'] for x in list_response.data_bundles]
-        print(ids)
+        logger.info(ids)
 
         # Find a DataBundle by alias
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 5]
```diff
@@ -1,660 +1,382 @@
...
-        except Exception as e:
-            print('The object no longer exists, '
-                  '404 not found. {}'.format(e))
+        logger.info("..........Delete the Bundle...............")
+        delete_response = self.request('DeleteDataBundle', data_bundle_id=data_bundle.id)
+        logger.info(delete_response.data_bundle_id)
+        with self.assertRaises(bravado.exception.HTTPNotFound):
+            self.request('GetDataBundle', data_bundle_id=update_response['data_bundle_id'])
 
         # Page through a listing of Data Bundles
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 6]
```diff
@@ -1,660 +1,382 @@
...
+        logger.info("..........Get all Objects in a Bundle..............")
+        data_bundle = self.request('GetDataBundle', data_bundle_id=data_bundle.id).data_bundle
         bundle_objects = []
         for data_object_id in data_bundle.data_object_ids:
             bundle_objects.append(self._client.GetDataObject(
                 data_object_id=data_object_id).result().data_object)
-        print([x.name for x in bundle_objects])
+        logger.info([x.name for x in bundle_objects])
 
         # DeleteDataBundle
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 7]
```diff
@@ -1,660 +1,382 @@
...
-        data_object = self._client.GetDataObject(
-            data_object_id=data_bundle.data_object_ids[0])\
-            .result().data_object
-        print(data_object.urls)
+        logger.info("..........Get an Object in a Bundle..............")
+        data_bundle = self.request('GetDataBundle', data_bundle_id=data_bundle.id).data_bundle
+        data_object = self.request('GetDataObject', data_object_id=data_bundle.data_object_ids[0]).data_object
+        logger.info(data_object.urls)
 
         # Get all DataObjects from a bundle
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 8]
```diff
@@ -1,660 +1,382 @@
...
         # Get all versions of a DataBundle
-        print("..........Get all Versions of a Bundle...............")
-        versions_response = self._client.GetDataBundleVersions(
-            data_bundle_id=data_bundle.id).result()
-        print(len(versions_response.data_bundles))
+        logger.info("..........Get all Versions of a Bundle...............")
+        versions_response = self.request('GetDataBundleVersions', data_bundle_id=data_bundle.id)
+        logger.info(len(versions_response.data_bundles))
 
         # Get a DataObject from a bundle
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 9]
```diff
@@ -1,660 +1,382 @@
...
-            checksum=list_request.checksum,
-            checksum_type=list_request.checksum_type,
-            page_size=list_request.page_size,
-            page_token=list_request.page_token).result()
-        print(len(list_response.data_bundles))
+        logger.info("..........List Data Bundles...............")
+        list_response = self.request('ListDataBundles')
+        logger.info(len(list_response.data_bundles))
 
         # Get all versions of a DataBundle
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 10]
```diff
@@ -1,660 +1,382 @@
...
+        updated_bundle = self.request('GetDataBundle', data_bundle_id=update_response['data_bundle_id']).data_bundle
+        logger.info('updated_bundle.aliases: %r', updated_bundle.aliases)
+        logger.info('updated_bundle.updated: %r', updated_bundle.updated)
+        logger.info('data_bundle.aliases: %r', data_bundle.aliases)
+        logger.info('data_bundle.updated: %r', data_bundle.updated)
+        # logger.info(updated_bundle.version)
+        # logger.info(updated_bundle.aliases)
+        self.assertEqual(updated_bundle.aliases[0], 'ghi')
 
         # ListDataBundles
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 11]
```diff
@@ -1,660 +1,382 @@
...
+        update_response = self.request('UpdateDataBundle', data_bundle=update_data_bundle,
+                                       query={'data_bundle_id': data_bundle.id})
+        logger.info("..........Get that Bundle.................")
+        updated_bundle = self.request('GetDataBundle', data_bundle_id=update_response['data_bundle_id']).data_bundle
+        logger.info('updated_bundle.aliases: %r', updated_bundle.aliases)
+        logger.info('updated_bundle.updated: %r', updated_bundle.updated)
+        logger.info('data_bundle.aliases: %r', data_bundle.aliases)
+        logger.info('data_bundle.updated: %r', data_bundle.updated)
+        # logger.info(updated_bundle.version)
+        # logger.info(updated_bundle.aliases)
```
Comment (achave11) [inline python/test/test_server.py]: https://github.com/ga4gh/data-object-service-schemas/blob/72b2c18d9bbfd68cea71eb9e79ee3819ca13c270/python/test/test_server.py#L318

Only leave this if they will be put on in the future, and add a TODO note.

File: python/test/test_server.py [hunk 12]
```diff
@@ -1,660 +1,382 @@
...
+            self.request('GetDataObject', data_object_id=id_)
+
+        # Mint a data bundle with the data objects we just created
+        data_bundle = self.generate_data_bundle(data_object_ids=ids)
+        self.request('CreateDataBundle', data_bundle=data_bundle)
+        # Does the data bundle we just created exist?
+        server_bdl = self.request('GetDataBundle', data_bundle_id=data_bundle.id).data_bundle
+        self.assertSameDataBundle(server_bdl, data_bundle)
 
         # UpdateDataBundle
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 13]
```diff
@@ -1,660 +1,382 @@
...
+            aliases.append(data_obj.aliases[0])
+            self.request('CreateDataObject', data_object=data_obj)
+        # Make sure that the data objects we just created exist
+        for id_ in ids:
+            self.request('GetDataObject', data_object_id=id_)
+
+        # Mint a data bundle with the data objects we just created
+        data_bundle = self.generate_data_bundle(data_object_ids=ids)
+        self.request('CreateDataBundle', data_bundle=data_bundle)
+        # Does the data bundle we just created exist?
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 14]
```diff
@@ -1,660 +1,382 @@
...
+            data_obj = self.generate_data_object()
+            ids.append(data_obj.id)
+            names.append(data_obj.name)
+            aliases.append(data_obj.aliases[0])
+            self.request('CreateDataObject', data_object=data_obj)
+        # Make sure that the data objects we just created exist
+        for id_ in ids:
+            self.request('GetDataObject', data_object_id=id_)
+
+        # Mint a data bundle with the data objects we just created
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 15]
```diff
@@ -1,660 +1,382 @@
...
-            data_bundle_id=data_bundle_id).result()
-        data_bundle = get_bundle_response.data_bundle
-        print(data_bundle)
-        print(data_bundle.id)
+            data_obj = self.generate_data_object()
+            ids.append(data_obj.id)
+            names.append(data_obj.name)
+            aliases.append(data_obj.aliases[0])
+            self.request('CreateDataObject', data_object=data_obj)
+        # Make sure that the data objects we just created exist
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 16]
```diff
@@ -1,660 +1,382 @@
...
-        URL = self._models.get_model('URL')
-        CreateDataObjectRequest = self._models.get_model(
-            'CreateDataObjectRequest')
-        DataObject = self._models.get_model(
-            'CreateDataObjectRequest')
-        ListDataObjectsRequest = self._models.get_model(
-            'ListDataObjectsRequest')
-
-        print("..........Create some data objects ............")
+        # Create data objects to populate the data bundle with
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 17]
```diff
@@ -1,660 +1,382 @@
...
+        self.request('CreateDataObject', data_object=data_obj)
+        # Get all of its versions. There should only be one - the original version
+        r = self.request('GetDataObjectVersions', data_object_id=data_obj.id)
+        self.assertEqual(len(r['data_objects']), 1)
+        # Now make a new version and upload it
+        data_obj.version = 'great-version'
+        data_obj.name = 'greatest-change'  # technically unnecessary, but just in case
+        self.request('UpdateDataObject', data_object=data_obj,
+                     query={'data_object_id': data_obj.id})
+        # If we check, there should now be two versions
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 18]
```diff
@@ -1,660 +1,382 @@
...
+        # :meth:`ga4gh.dos.test.ComplianceTest.test_list_data_object_querying`.
+
+    def test_data_object_versions(self):
+        # Create a data object
+        data_obj = self.generate_data_object()
+        self.request('CreateDataObject', data_object=data_obj)
+        # Get all of its versions. There should only be one - the original version
+        r = self.request('GetDataObjectVersions', data_object_id=data_obj.id)
+        self.assertEqual(len(r['data_objects']), 1)
+        # Now make a new version and upload it
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 19]
```diff
@@ -1,660 +1,382 @@
...
+        results = self.request('ListDataObjects', query={'url': data_obj.urls[0].url})
+        self.assertEqual(len(results['data_objects']), 1)
+        # The more advanced ListDataObjects testing is left to
+        # :meth:`ga4gh.dos.test.ComplianceTest.test_list_data_object_querying`.
+
+    def test_data_object_versions(self):
+        # Create a data object
+        data_obj = self.generate_data_object()
+        self.request('CreateDataObject', data_object=data_obj)
+        # Get all of its versions. There should only be one - the original version
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 20]
```diff
@@ -1,660 +1,382 @@
...
+                                      'checksum_type': data_obj.checksums[0].type})
+        self.assertEqual(len(results['data_objects']), 1)
+        # and by a unique url.
+        results = self.request('ListDataObjects', query={'url': data_obj.urls[0].url})
+        self.assertEqual(len(results['data_objects']), 1)
+        # The more advanced ListDataObjects testing is left to
+        # :meth:`ga4gh.dos.test.ComplianceTest.test_list_data_object_querying`.
+
+    def test_data_object_versions(self):
+        # Create a data object
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 21]
```diff
@@ -1,660 +1,382 @@
...
+        # by a unique checksum, ...
+        results = self.request('ListDataObjects',
+                               query={'checksum': data_obj.checksums[0].checksum,
+                                      'checksum_type': data_obj.checksums[0].type})
+        self.assertEqual(len(results['data_objects']), 1)
+        # and by a unique url.
+        results = self.request('ListDataObjects', query={'url': data_obj.urls[0].url})
+        self.assertEqual(len(results['data_objects']), 1)
+        # The more advanced ListDataObjects testing is left to
+        # :meth:`ga4gh.dos.test.ComplianceTest.test_list_data_object_querying`.
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 22]
```diff
@@ -1,660 +1,382 @@
...
+        self.request('CreateDataObject', data_object=data_obj)
+        # We should be able to retrieve the data object by a unique alias, ...
+        results = self.request('ListDataObjects', query={'alias': data_obj.aliases[0]})
+        self.assertEqual(len(results['data_objects']), 1)
+        # by a unique checksum, ...
+        results = self.request('ListDataObjects',
+                               query={'checksum': data_obj.checksums[0].checksum,
+                                      'checksum_type': data_obj.checksums[0].type})
+        self.assertEqual(len(results['data_objects']), 1)
+        # and by a unique url.
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 23]
```diff
@@ -1,660 +1,382 @@
...
+        self.assertEqual(ctx.exception.status_code, 404)
+
+    def test_list_data_object_querying(self):
+        # Create a data object
+        data_obj = self.generate_data_object()
+        self.request('CreateDataObject', data_object=data_obj)
+        # We should be able to retrieve the data object by a unique alias, ...
+        results = self.request('ListDataObjects', query={'alias': data_obj.aliases[0]})
+        self.assertEqual(len(results['data_objects']), 1)
+        # by a unique checksum, ...
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 24]
```diff
@@ -1,660 +1,382 @@
...
+        # Make sure it's gone
+        with self.assertRaises(bravado.exception.HTTPNotFound) as ctx:
+            self.request('GetDataObject', data_object_id=data_obj.id)
+        self.assertEqual(ctx.exception.status_code, 404)
+
+    def test_list_data_object_querying(self):
+        # Create a data object
+        data_obj = self.generate_data_object()
+        self.request('CreateDataObject', data_object=data_obj)
+        # We should be able to retrieve the data object by a unique alias, ...
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 25]
```diff
@@ -1,660 +1,382 @@
...
+        self.assertSameDataObject(data_obj, response.data_object)
+        # Begone foul data object
+        self.request('DeleteDataObject', data_object_id=data_obj.id)
+        # Make sure it's gone
+        with self.assertRaises(bravado.exception.HTTPNotFound) as ctx:
+            self.request('GetDataObject', data_object_id=data_obj.id)
+        self.assertEqual(ctx.exception.status_code, 404)
+
+    def test_list_data_object_querying(self):
+        # Create a data object
```
Comment (achave11) [inline python/test/test_server.py]: Nit: single line comment. Or remove.

File: python/test/test_server.py [hunk 26]
```diff
@@ -1,660 +1,382 @@
...
+    #     self.request('UpdateDataObject', data_object=data_obj_2,
+    #                  query={'data_object_id': data_obj_1.id})
+    #     # Try to retrieve the data object by its new id and its old id
+    #     # The former should succeed:
+    #     response = self.request('GetDataObject', data_object_id=data_obj_2.id)
+    #     self.assertSameDataObject(response.data_object, data_obj_2)
+    #     # And the latter should fail:
+    #     with self.assertRaises(bravado.exception.HTTPNotFound) as ctx:
+    #         self.request('GetDataObject', data_object_id=data_obj_1.id)
+    #     self.assertEqual(ctx.exception.status_code, 404)
```
Comment (achave11) [inline python/test/test_server.py]: Only leave this if they will be put on in the future, and add a TODO note.

**Comment by achave11** [comment]
I will approve if you remove the commented out code, or if you add the `#TODO` convention if you are trying to use the code in the future.

**Comment by natanlao** [comment]
Thanks for the review, @achave11!

**Pull request merged**